fuzz/do_fuzz.py

```diff
@@ -0,0 +1,129 @@
+        return input_data
+    except Exception:
+        raise TypeError
+
+
+def generate_int(data_format):
+    is_signed = data_format['is_signed']
+    width = data_format['width']
+
+    if is_signed:
+        result = random.randrange(-2 ** (width - 1) + 1, 2 ** (width - 1) - 1)
+    else:
+        result = random.randrange(0, 2 ** width - 1)
+    return result
+
+
+def construct_json(data_list, obj, template):
+    with open(template) as json_file:
+        data = json.load(json_file)
+    data[obj]["data"] = data_list
+
+    with open(template, "w") as json_file:
+        json.dump(data, json_file)
+
+
+def compare_object(a, b):
+    if type(a) != type(b):
+        return False
+    elif type(a) is dict:
+        return compare_dict(a, b)
+    elif type(a) is list:
+        return compare_list(a, b)
+    else:
+        return a == b
+
+
+def compare_dict(a, b):
+    if len(a) != len(b):
+        return False
+    else:
+        for k, v in a.items():
+            if k not in b:
+                return False
+            else:
+                if not compare_object(v, b[k]):
+                    return False
+    return True
+
+
+def compare_list(a, b):
+    if len(a) != len(b):
+        return False
+    else:
+        for i in range(len(a)):
+            if not compare_object(a[i], b[i]):
+                return False
+    return True
+
+
+def do_fuzz(prog, spec, template, iteration):
+    try:
+        prog_name = os.path.splitext(prog)[0]
+        prog_file_type = os.path.splitext(prog)[1]
+        spec_name = os.path.splitext(spec)[0]
+        spec_file_type = os.path.splitext(spec)[1]
+        if prog_file_type != ".futil":
+            prog = prog_name + ".futil"
+            os.system(f"fud e {prog} --to futil > {prog}")
+
+        elif spec_file_type != ".futil":
+            spec = spec_name + ".futil"
+            os.system(f"fud e {spec} --to futil > {spec}")
+
+        for i in range(iteration):
+            generate_data(template)
+            if os.stat(template).st_size != 0:
+                os.system(
+                    f"fud e {prog} -s verilog.data {template} --to dat -q --through icarus-verilog > result1.json")
+                os.system(
+                    f"fud e {spec} -s verilog.data {template} --to dat -q --through icarus-verilog > result2.json")
+                with open('result1.json', 'r') as f1, open('result2.json', 'r') as f2:
+                    f1_result = json.load(f1)
+                    f2_result = json.load(f2)
+                    obj1 = f1_result['memories']
+                    obj2 = f2_result['memories']
+                    assert compare_object(obj1, obj2), "Failed with unequal output!"
+    except Exception as e:
+        logging.error(e)
+        logging.error(f'fail unequal output for prog: {prog}, spec: {spec}')
+
+
+if __name__ == "__main__":
+    prog = sys.argv[1]
+    spec = sys.argv[2]
+    template = sys.argv[3]
+    iteration = sys.argv[4]
```

Comment:
Should I create a README to write out the example cli invocation?

This is an example: `python do_fuzz.py std_sdiv.futil std_div.futil template.json 2`, where we compare the signed and unsigned division in calyx, with a template on how these two calyx files read input, and we run fuzzing twice.

Here is a template file example for division:
`{
  "left": {
    "data": [0, 0, 0, 0],
    "format": {
      "numeric_type": "bitnum",
      "is_signed": true,
      "width": 32
    }
  },
  "right": {
    "data": [0, 0, 0, 0],
    "format": {
      "numeric_type": "bitnum",
      "is_signed": true,
      "width": 32
    }
  },
  "out_rem": {
    "data": [0, 0, 0, 0],
    "format": {
      "numeric_type": "bitnum",
      "is_signed": true,
      "width": 32
    }
  },
  "out_quot": {
    "data": [0, 0, 0, 0],
    "format": {
      "numeric_type": "bitnum",
      "is_signed": true,
      "width": 32
    }
  }
}`

Reply:
Yeah, I think you can add a README in the `fuzz` folder with this example and it would be a good place to start. Eventually, we can move it into the Calyx documentation website.